### PR TITLE
Return input argument if no GPUs werde detected in get_gpu

### PR DESCRIPTION
**Pull request by masterleinad**
Fixes https://github.com/kokkos/kokkos/issues/4621. @jczhang07 Can you please take a look if this works for you?

Essentially, the changes here are only relevant for the SYCL backend. If a specific device is requested (id>=0) we still error out if there are no GPUs. If that is not the case (id<0), we use the default selector to run on a host queue if no GPUs are detected. This (intended) behavior is not changed here but we have `use_gpu` return the input argument when no GPU is detected (instead of assigning GPU 0 which doesn't exist in that case).

### CODE REVIEW

**Review by dalg24 — CHANGES_REQUESTED** [review]
This makes that code even less readable

**Comment by masterleinad** [comment]
> This makes that code even less readable

What is it specifically that you want me to change? Are you criticizing the intended behavior or do you merely want to change the implementation to be more readable? 

In the latter case, we could copy
```
return sycl::device::get_devices(sycl::info::device_type::gpu).size();
```
before calling `get_gpu` in the SYCL case (although this function is then called twice for the regular case of running on a GPU). Would that make you accept the changes?

**Comment by jczhang07** [comment]
@masterleinad The fix works for me. Thanks.

**Comment by masterleinad** [comment]
@dalg24 I avoided modifying `get_gpu` now but moved the logic to the respective `SYCL` call site.

**Review by dalg24 — APPROVED** [review]

**Comment by masterleinad** [comment]
Retest this please.

**Comment by jczhang07** [comment]
> Retest this please.
@masterleinad Reested and it worked. Thanks.

**Review by Rombur — CHANGES_REQUESTED** [review]

**Review by Rombur — APPROVED** [review]

File: core/src/SYCL/Kokkos_SYCL.cpp [hunk 1]
```diff
@@ -284,7 +284,11 @@ int g_sycl_space_factory_initialized =
     Kokkos::Impl::initialize_space_factory<SYCLSpaceInitializer>("170_SYCL");
 
 void SYCLSpaceInitializer::initialize(const InitArguments& args) {
-  int use_gpu = Kokkos::Impl::get_gpu(args);
+  // If there are no GPUs return whatever else we can run on if no specific GPU is
+  // requested.
```
Comment (dalg24) [inline core/src/SYCL/Kokkos_SYCL.cpp:288]:
```suggestion
  // If there are no GPUs return whatever else we can run on if no specific GPU
  // is requested.
```

**Comment by dalg24** [comment]
Retest this please

**Comment by jczhang07** [comment]
> Retest this please

Works here.

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
> Works here.

@jczhang07 The comments are not for you but for our CI. :slightly_smiling_face:

**Pull request merged**